### ISSUE TITLE
Docker issues

### ISSUE THREAD

**Issue by dmikushin** [mjl-/mox#35]
Mox is an interesting software. At first I thought it's a simple pet project, but actually it's quite smart!

I only have some concerns about the Docker setup. I think the instructions provided in `docker-compose.yml` cannot work. First, there is no `mox` user within the container, and creating it there must be a part of `Dockerfile`. In principle, it would not be a big problem to do `docker-compose run mox mox quickstart you@yourdomain.example root`, since this root belongs to the container, right?

Another issue is with ports. Mox container attaches to a host network and tries to bind the priviledged ports. I think it's not possible without `priviledged: true` mode added into `docker-compose.yml`, do you agree?

Thanks for your great work!

**Comment by mjl-** [comment]
hi, glad to hear you find mox interesting!

about docker-compose with the quickstart: the goal of the suggested command is to run the quickstart with a uid as last parameter, not a user name. by using a uid, there is no need for the user to exist in the docker container (if i had to create one in the docker container, i wouldn't know which uid to pick, so i'm leaving it to the admin setting up mox). mox will just start mox with the uid, skipping a lookup of username to uid. that's `docker-compose run mox mox quickstart you@yourdomain.example $(id -u mox)`. the `$(id -u mox)` is evaluated in the host operating system, which will have a "mox" user after running `useradd -d $PWD mox`. though you can also just make up a uid and not create mox user on the host OS. i would recommend against passing "root" as username, it would try to run the whole process with root privileges.

as for the ports: mox starts as root, binds to the "privileged" ports, then starts a process with the unprivileged uid and passes the open network listenening sockets to that unprivileged process. this is all "regular" (unix and) docker use, i don't think "privileged: true" is needed. as far as i know, with "privileged: true" there is no isolation of the docker container from the host operating system. it seems to be used to run docker containers from within docker. i don't think docker's "privileged: true" refers to listening on "privileged" ports.

in short, it seems your two suggestions would remove two pretty important security isolation mechanisms.

did you try the commands and get an error message? if so, could you open this issue again with the error message? that should help pinpoint the problem. thanks for taking the time to create this issue!

**Comment by dmikushin** [comment]
Hi @mjl- , thank you for explaining your thoughts about this issue.

1. Port 25 of host is a privileged port. It is not possible to bind it without extra permissions, if it is binding a port belonging to the host machine. Since the very early day of security settings, ports <1024 have restricted binding rights. So, any program (including Docker), which tries to work with these ports needs to elevate the privileges. Below is an example error as you asked: 

```
mox_1  | l=fatal m="smtp: listen for smtp" err="listen tcp6 [...]:25: bind: cannot assign requested address" pkg=smtpserver protocol=smtp listener=public
```

2. This is correct

> that's `docker-compose run mox mox quickstart you@yourdomain.example $(id -u mox)`. the $(id -u mox) is evaluated in the host operating system, which will have a "mox" user after running useradd -d $PWD mox. though you can also just make up a uid and not create mox user on the host OS. 

but again makes no sense with respect to how Docker works. UID belongs to a host system, but does not exist in the Docker container. Mox should be running inside the docker container, but it does not have a user on behalf of which to drop the privileges.

Despite an excellent work on mox, you don't understand how Docker works. I think the main deployment scenario of Mox is a host OS installation. Perhaps, Docker scenario was requested by other users, but was never developed enough to a working state.

PS Please don't close issues prematurely, it's an unwelcoming practice.

**Comment by mjl-** [comment]
Thanks for these details and the error message!

I believe that the error message "cannot assign requested address" indicates that the IP address that mox is trying to listen on, is not available inside the docker container. If the process wouldn't have permission to listen on the privileged port, the error message would be "permission denied". The following program and executions illustrate that (ran as regular user, i.e. without permission to bind to privileged ports):

```
$ cat main.go
package main

import (
        "net"
        "log"
        "os"
)

func main() {
        log.SetFlags(0)
        addr := os.Args[1]
        if _, err := net.Listen("tcp", addr); err != nil {
                log.Fatalf("listen %v: %v", addr, err)
        }
        log.Print("listening on %v", addr)
}
```

In this first execution, 1.2.3.4 is not an IP on my machine:

```
$ go run main.go 1.2.3.4:25
listen 1.2.3.4:25: listen tcp 1.2.3.4:25: bind: cannot assign requested address
exit status 1
```

In this second execution, I don't have permission to listen on port 25:

```
$ go run main.go 192.168.1.10:25
listen 192.168.1.185:10: listen tcp 192.168.1.10:25: bind: permission denied
exit status 1
```

Mox is starting up as root, and binds to its network ports as root, so does have permission to listen on port 25. You could try running this inside a docker container (as root, and with a valid IP, or just `:25`).

As for the issue with the UID. I think I know enough about docker and linux/unix to understand how it handles uids (though of course I can be wrong or my understanding outdated). I hope I can explain my understanding properly.

By default, uids as used in docker are the same uids as used on the host OS. This is important for file system operations: If a host OS directory is mounted (bound) in a container, and a process inside that container that is running as uid $X (e.g. 1234) creates a file in that mounted directory, the uid (owner) of the file is uid $X. If you look at the file on the host OS, you'll see that same uid $X. A typical docker-compose invocation/yml doesn't do anything about/with uids. By default, processes are started as root, and as such can start new processes under any other uid, and create files under any other uid.

You don't need to create a username to use a uid. Linux/unix kernels and their API's, don't even know about user names, only about uids. Usernames are just mappings of names to uids, convenient for humans, stored in /etc/passwd. The problem here is that I don't want to run mox inside the docker container as root. But which uid should it be if not 0? [1] It is relevant to the host OS, because files are going to show up in the host OS file system with the chosen uid. That's why I leave it to the host OS admin to pick a uid. In mox's docker-compose.yml I'm suggesting invoking "useradd" on the host OS so it assigns its next free/unused uid. Mox must run as that uid, that's why the suggested quickstart command uses "id -u mox" (running on the host OS). The quickstart will write that uid to config/mox.conf, so that the actual mox process in the container (started as root) can switch to that uid. Files created by mox will show up in the host OS file system under that same uid. There is no need for that same user to "exist" (have an entry in /etc/passwd) inside the docker container. The unix/Go API's work with uids, and the programmer typically converts user names to uids explicitly by doing lookups. It wouldn't _hurt_ to have a username for the chosen uid inside the container, but I wanted to keep the instructions (and code) simple.

Perhaps we're looking at it slightly differently. If you assume docker containers are fully isolated from each other, the value of the non-root uid that mox runs under doesn't matter, and could be anything other than 0 (don't use 0: it has special powers in many permission checks in the kernel). In practice, with a standard docker daemon, the uids leak through to the host OS. So when looking at it from the host OS admin, the specific uid does matter. In my experience, docker has plenty of surprising differences between what people (including me) think/hope it delivers (especially with respect to isolation) and what it actually delivers.

I closed the issue mostly because I thought I understood there wasn't an actual problem, which was in part because there was no error message. Clearly a bit too eager, sorry about that. I think you can reopen issues on github, so please feel free to do so. If you know of a better way to deal with sharing a uid between host OS and container,  especially assigning a unique uid on the host OS to a container/service, I would be happy to learn more!

[1] I've seen docker images pick some (random?) uid, for which they created a username inside the image. I think that's not a good idea: the uid makes it into the host OS file system, and that same uid may already "belong" to an existing user on the host OS. That existing host OS user now has access to the files of the (presumed isolated) container. I've always been surprised (flabbergasted?) that docker seemed to leave this problem unsolved (similar to how I was surprised that the default behavior of docker containers is to start as root and how people are happy to run docker images off the internet!).

**Comment by dmikushin** [comment]
You are right, the issue is even deeper than I thought. Note `tcp6` in my error message: Docker even does not support IPv6 networking by default. In order to bind tcp6, some special settings need to be applied. Please don't get me wrong: a lot of work needs to be done before this container could actually be deployed.

**Comment by mjl-** [comment]
If mox is trying to listen on ipv6 when that isn't available, it sounds like a mismatch between the IPs that the quickstart generated vs what is actually in the docker container. Was the Listen IP address that caused the trouble "::" and/or "::1"? Or was it more an actual host IPv6 address? The ipv6 any and localhost addresses, along with similar IPv4 addresses, are used in the config if no IPs could be found otherwise. I could see how such a mismatch could occur when the quickstart command was run in a different way (e.g. not through the docker-compose file with host networking) than the actual container. If that is the reason for the mismatch, perhaps we can improve the docs around that. If there is another reason for the mismatch, it would be good to try to address that.
